### PR TITLE
Add favorite colorway

### LINKED ISSUE #476: Favorite colorway
Because we address the favorite keyboard in: https://github.com/qmk/qmk_configurator/pull/465

Why not adding a favorite colorway in the app?

### PR DESCRIPTION
**Pull request by zekth**
fix #476 

This is not the cleanest way because the colorway handling is not that clean. It works like a charm now and i think we can refactor this part when we will redesign the app.
Screenshot 2019-11-17 at 11 42 38

### CODE REVIEW

**Comment by noroadsleft** [comment]
This doesn't load for me at all.

image



```text
13:17:06.926 [Vue warn]: Error in render: "TypeError: this.configuratorSettings.favoriteColor.toLowerCase is not a function"

found in

---> <Main> at src/components/Main.vue
       <Configurator> at src/views/Home.vue
         <App> at src/App.vue
           <Root> vue.runtime.esm.js:619
13:17:06.929 TypeError: "this.configuratorSettings.favoriteColor.toLowerCase is not a function"
    isFavoriteColor Main.vue:122
    VueJS 4
    render Main.vue:108
    VueJS 45
    <anonymous> main.js:93
    js app.js:9899
    __webpack_require__ app.js:724
    fn app.js:101
    1 app.js:10320
    __webpack_require__ app.js:724
    <anonymous> app.js:791
    <anonymous> app.js:794
vue.runtime.esm.js:1888
13:17:07.046 [Vue warn]: Error in mounted hook: "TypeError: this.configuratorSettings.favoriteColor.toLowerCase is not a function"

found in

---> <Main> at src/components/Main.vue
       <Configurator> at src/views/Home.vue
         <App> at src/App.vue
           <Root> vue.runtime.esm.js:619
    VueJS 15
    <anonymous> main.js:93
    js app.js:9899
    __webpack_require__ app.js:724
    fn app.js:101
    1 app.js:10320
    __webpack_require__ app.js:724
    <anonymous> app.js:791
    <anonymous> app.js:794
13:17:07.047 TypeError: "this.configuratorSettings.favoriteColor.toLowerCase is not a function"
    mounted Main.vue:145
    VueJS 11
    <anonymous> main.js:93
    js app.js:9899
    __webpack_require__ app.js:724
    fn app.js:101
    1 app.js:10320
    __webpack_require__ app.js:724
    <anonymous> app.js:791
    <anonymous> app.js:794
vue.runtime.esm.js:1888

```

**Comment by zekth** [comment]
Yup i've seen that with the CI
fixing it

**Comment by noroadsleft** [comment]
Works now, but there's no visual feedback in real time.

**Comment by noroadsleft** [comment]
Can we apply `cursor: pointer;` to `#favorite-colorway` in CSS, like the icon for favorite keyboard?

**Comment by zekth** [comment]
I've fixed some issues. Now it's properly working

File: src/components/Main.vue [hunk 1]
```diff
@@ -39,6 +39,15 @@
               >{{ name }}</option
             >
           </select>
+          <a
+            id="favorite-colorway"
+            v-on:click="favColor"
```
Comment (yanfali) [inline src/components/Main.vue]: use @ not v-on

File: src/components/Main.vue [hunk 2]
```diff
@@ -39,6 +39,15 @@
...
+            v-bind:class="{
```
Comment (yanfali) [inline src/components/Main.vue]: :class not v-bind

File: src/components/Main.vue [hunk 3]
```diff
@@ -107,14 +117,40 @@ export default {
...
+      } else {
+        this.setFavoriteColor(this.displayColorways[this.curIndex]);
+      }
+    }
   },
   mounted() {
     jquery.init();
+    // Loading favorite color
+    if (this.configuratorSettings.favoriteColor) {
+      for (let i = 0; i < this.displayColorways.length; i++) {
```
Comment (yanfali) [inline src/components/Main.vue]:
```javascript
this.curIndex = this.displayColorways.find(color => color.toLowerCase === this.configuratorSettings.favoriteColor.toLowerCase);
```

Comment (yanfali) [reply src/components/Main.vue]: that should get rid of a for loop

File: src/scss/themes.scss [hunk 4]
```diff
@@ -39,12 +39,17 @@ html {
...
     }
-    #favorite-keyboard {
+    #favorite-keyboard,
+    #favorite-colorway {
       color: grey;
       &.active {
         color: $qmk-green;
       }
     }
+    #favorite-colorway {
```
Comment (yanfali) [inline src/scss/themes.scss]: favorite-keyboard has pointer behavior already?

Comment (yanfali) [reply src/scss/themes.scss]: does this belong in themes or in styles.scss?

File: src/components/Main.vue (hunk 3 above)
Comment (zekth) [reply src/components/Main.vue]: indeed

File: src/scss/themes.scss (hunk 4 above)
Comment (zekth) [reply src/scss/themes.scss]: Oh yeah completely mixed those sheets

**Comment by zekth** [comment]
> [10f21e5](https://github.com/qmk/qmk_configurator/commit/10f21e51238e15f870a9fbf0577e66fbbe16eda6) is non-working for me.
> 
> ```
> 23:55:42.626 [Vue warn]: Error in render: "TypeError: state.colorways[state.colorwayIndex] is undefined"
> 
> found in
> 
> ---> <VisualKeymap> at src/components/VisualKeymap.vue
>        <Main> at src/components/Main.vue
>          <Configurator> at src/views/Home.vue
>            <App> at src/App.vue
>              <Root> vue.runtime.esm.js:619
> 23:55:42.626 TypeError: "state.colorways[state.colorwayIndex] is undefined"
>     colorway keymap.js:30
>     wrappedGetter vuex.esm.js:762
>     partial vuex.esm.js:95
>     VueJS 3
>     get vuex.esm.js:564
>     mappedGetter vuex.esm.js:900
>     VueJS 3
>     currentLayer VisualKeymap.vue:85
>     VueJS 4
>     render VisualKeymap.vue:9
>     VueJS 7
> vue.runtime.esm.js:1888
>     VueJS 10
> ```

Tried to reproduce. Can't. Could you share your localstorage please?

**Comment by yanfali** [comment]
@zekth you removed the guard. You have to put that back in.
```
 if (this.configuratorSettings.favoriteColor) {
```

**Comment by yanfali** [comment]
@zekth can you add a title hover with help text explaining what the star is please?

**Comment by yanfali** [comment]
same thing for the star next to favorite keyboard please

**Comment by noroadsleft** [comment]
> Tried to reproduce. Can't. Could you share your localstorage please?

```
configuratorSettings:"{"version":1,"darkmodeEnabled":false,"favoriteKeyboard":"","favoriteColor":""}"
loglevel:webpack-dev-server:"SILENT"
```

It works if I set `favoriteColor` to a valid value, but without a colorway set as favorite it errors.

**Pull request merged**